File: solutions/scala/src/test/scala/CalculatorTests.scala
import org.scalatest.TryValues.convertTryToSuccessOrFailure
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class CalculatorTests
  extends AnyFunSuite {

  test("calculator should support add") {
    // Arrange
    val calculator = new Calculator()

    // Act
    val result = calculator.calculate(9, 3, operator = Calculator.add)

    // Assert
    assert(result.success.value == 12)
  }

  test("calculator should support multiply") {
    // Arrange
    val calculator = new Calculator()

    // Act
    val result = calculator.calculate(3, 76, operator = Calculator.multiply)

    // Assert
    assert(result.success.value == 228)
  }

  test("calculator should support divide") {
    // Arrange
    val calculator = new Calculator()

    // Act
    val result = calculator.calculate(9, 3, operator = Calculator.divide)

    // Assert
    assert(result.success.value == 3)
  }

  test("calculator should support subtract") {
    // Arrange
    val calculator = new Calculator()

    // Act
    val result = calculator.calculate(9, 3, operator = Calculator.subtract)

    // Assert
    assert(result.success.value == 6)
  }

  test("calculator should fail when operator not supported") {
    val calculator = new Calculator()
    assert(
      calculator
        .calculate(9, 3, operator = "UnsupportedOperator")
        .failure
        .exception
        .getMessage == "Not supported operator"
    )
  }
}

File: solutions/java/src/test/java/CalculatorTests.java
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import unit.tests.intro.Calculator;

import static org.junit.jupiter.api.Assertions.*;

class CalculatorTests {
    @Test
    void supportAdd()
    {
        // Arrange
        var calculator = new Calculator();

        // Act
        var result = calculator.calculate(9, 3, Calculator.ADD);

        // Assert
        assertEquals(12, result);
    }


    @Test
    void supportMultiply()
    {
        // Arrange
        var calculator = new Calculator();

        // Act
        var result = calculator.calculate(3, 76, Calculator.MULTIPLY);

        // Assert
        assertEquals(228, result);
    }

    @Test
    void supportDivide()
    {
        // Arrange
        var calculator = new Calculator();

        // Act
        var result = calculator.calculate(9, 3, Calculator.DIVIDE);

        // Assert
        assertEquals(3, result);
    }

    @Test
    void supportSubtract()
    {
        // Arrange
        var calculator = new Calculator();

        // Act
        var result = calculator.calculate(9, 3, Calculator.SUBTRACT);

        // Assert
        assertEquals(6, result);
    }

    @Test
    void failWhenOperatorNotSupported()
    {
        var calculator = new Calculator();
        var exception = Assertions.assertThrows(IllegalArgumentException.class, () -> calculator.calculate(9, 3, "UnsupportedOperator"));
        assertEquals("Not supported operator", exception.getMessage());
    }
}